django/utils/functional.py

```diff
@@ -464,3 +464,15 @@ def partition(predicate, values):
     for item in values:
         results[predicate(item)].append(item)
     return results
+
+
+def flatten(x):
+    """
+    Create a generator that returns all the elements of the (nested) array. e.g.:
+
+         >>> list(flatten([[1, 2], 3, [[4, 5], [6, 7]]]))
+             [0, 1, 2, 3, 4, 5, 6, 7]
+    """
+    if not isinstance(x, (tuple, list)):
+        return iter([x])
+    return itertools.chain.from_iterable((flatten(y) for y in x))
```

Comment:
Honestly I think the other version is more readable. This version may have the possible slight performance benefit of being C code but that is just guessing and would need to be measured.

We may also need to include all iterables _except_ strings. 🤔

Given that an additional function in utils is accepted (it may very well be decided against) it might be worth asking around from others who as to what they think. One person I'm thinking of is @kezabelle as they've done some performance related refactoring recently ☺️

Reply:
@kezabelle if you do read this for context: we had a need for a list/tuple flattening function. I suggested perhaps adding to utils.

The previous version was something like:

```
def flatten(x):
    if not isinstance(x, (tuple, list)):
        return iter([x])
    return (z for for y in x for z in flatten(y))
```

Reply:
>but that is just guessing and would need to be measured.

Indeed!
On a mac OS 12.0.1, Apple M1.

The three functions
```python
def flatten_itertools_iter(x):
    to_visit = iter([x])
    while True:
        try:
            curr = next(to_visit)
        except StopIteration:
            return
        if not isinstance(curr, (tuple, list)):
            yield curr
        else:
            to_visit = itertools.chain(iter(curr), to_visit)


def flatten_itertools_rec(x):
    if not isinstance(x, (tuple, list)):
        return iter([x])
    return itertools.chain.from_iterable((flatten_rec(y) for y in x))

def flatten_rec(x):
    if not isinstance(x, (tuple, list)):
        return iter([x])
    return (z for y in x for z in flatten_rec(y))
```

```python

r = [[None for _ in range(10) for _ in range(365)]]


def test_flatten_rec(benchmark):
  benchmark(lambda: list(flatten_rec(r)))


def test_flatten_itertools_rec(benchmark):
  benchmark(lambda: list(flatten_itertools_rec(r)))


def test_flatten_itertools_iter(benchmark):
  benchmark(lambda: list(flatten_itertools_iter(r)))
```
The benchmark with pytest-benchmark
https://pypi.org/project/pytest-benchmark/
```
--------------------------------------------------------------------------------
Name (time in us)                    Min                 Max                Mean
--------------------------------------------------------------------------------
test_flatten_itertools_iter     365.4170 (1.0)      484.4580 (1.0)      370.2386
test_flatten_itertools_rec      600.7500 (1.64)     659.4590 (1.36)     609.8830
test_flatten_rec                654.3750 (1.79)     820.7500 (1.69)     660.6415
--------------------------------------------------------------------------------

```

I could send you the histogram if you'd like more info too.

@shangxiao @kezabelle WDYT?

Reply:
@yetanotherion Maybe inplace assignment is faster? Something like this here: https://stackoverflow.com/a/10824086/12548337

Reply:
Indeed, inplace assignment would be faster @jerch! I created new tests to evaluate the benefits of using a generator:

```python
def flatten_itertools_iter(x):
    to_visit = iter([x])
    while True:
        try:
            curr = next(to_visit)
        except StopIteration:
            return
        if not isinstance(curr, (tuple, list)):
            yield curr
        else:
            to_visit = itertools.chain(iter(curr), to_visit)

def flatten_itertools_rec(x):
    if not isinstance(x, (tuple, list)):
        return iter([x])
    return itertools.chain.from_iterable((flatten_rec(y) for y in x))

def flatten_rec(x):
    if not isinstance(x, (tuple, list)):
        return iter([x])
    return (z for y in x for z in flatten_rec(y))

def flatten_inplace(items, seqtypes=(list, tuple)):
    try:
        for i, x in enumerate(items):
            while isinstance(items[i], seqtypes):
                items[i:i+1] = items[i]
    except IndexError:
        pass
    return items

def flatten_iter(iterable):
    iterator, sentinel, stack = iter(iterable), object(), []
    while True:
        value = next(iterator, sentinel)
        if value is sentinel:
            if not stack:
                break
            iterator = stack.pop()
        elif isinstance(value, str):
            yield value
        else:
            try:
                new_iterator = iter(value)
            except TypeError:
                yield value
            else:
                stack.append(iterator)
                iterator = new_iterator
```
```
Name (time in ns)                             Min                       Max                      Mean
-----------------------------------------------------------------------------------------------------
test_flatten_iterools_iter_any           625.0000 (1.0)         14,000.0000 (5.34)           754.5814
test_flatten_iter_any                    720.8000 (1.15)         2,622.9000 (1.0)            734.5171
test_flatten_inplace_any             263,000.0000 (420.80)     335,250.0000 (127.82)     266,643.3846
test_flatten_inplace                 277,459.0000 (443.93)   4,862,167.0000 (>1000.0)    296,771.0054
test_flatten_itertools_iter          365,458.0000 (584.73)     849,250.0000 (323.78)     381,319.9589
test_flatten_itertools_rec           603,042.0000 (964.87)     714,833.0000 (272.54)     610,892.7198
test_flatten_rec                     654,708.0000 (>1000.0)    795,417.0000 (303.26)     662,324.5929
test_flatten_iter                  1,022,500.0000 (>1000.0)  1,082,125.0000 (412.57)   1,029,526.2549
-----------------------------------------------------------------------------------------------------
```

From a performance point of view, it seems that `flattern_itertools_iter` wins:
- faster version among the ones that output a generator,
- all generator versions, outperform the non generator implementation on the any use case.

Now the question seems to be maintainability wise, whether the added complexity is necessary.

Given that I wrote this version I'm very biased :) So I'd need your inputs to choose which version to write.
Once we have a consensus on which version to use, I'll push a new PR with the chosen implementation.

Thanks again 🙏 for your precious insights on this PR @jerch @shangxiao !

Reply:
Thank you so much @jerch, the numbers speak for themselves, that's a brilliant idea!
```
Name (time in ns)                            Min                       Max                    Mean
------------------------------------------------------------------------------------------------------------
test_flatten_unroll3_any                429.2000 (1.0)          3,100.0000 (2.10)         451.1658 (1.0)
test_flatten_itertools_iter_any         675.0000 (1.57)         1,479.1500 (1.0)          697.5274 (1.55)
test_flatten_iter_any                   752.0500 (1.75)         2,447.9500 (1.65)         779.7358 (1.73)
test_flatten_unroll3                238,375.0000 (555.39)     293,000.0000 (198.09)   243,921.7601 (540.65)
test_flatten_inplace_any            246,625.0000 (574.62)     288,500.0000 (195.04)   252,525.5564 (559.72)
test_flatten_inplace                266,250.0000 (620.34)     331,209.0000 (223.92)   275,429.1903 (610.48)
test_flatten_inplace_copy           281,541.0000 (655.97)     378,291.0000 (255.75)   282,962.4365 (627.18)
test_flatten_itertools_iter         368,750.0000 (859.16)     421,417.0000 (284.90)   377,882.8414 (837.57)
test_flatten_itertools_rec          607,000.0000 (>1000.0)    706,584.0000 (477.70)   621,654.6501 (>1000.0)
test_flatten_rec                    663,625.0000 (>1000.0)    791,834.0000 (535.33)   678,220.4457 (>1000.0)
test_flatten_iter                   971,375.0000 (>1000.0)  1,112,042.0000 (751.81)   996,490.5042 (>1000.0)
------------------------------------------------------------------------------------------------------------
```
WDYT @shangxiao?

Reply:
Oh wow thats even faster than inplace, did not expect this.

Well it comes with downsides:
- it is still recursive, thus recursion error still might occur
- it degrades for higher nested structures, as it is still not subcall free
- the seqtype test against `(tuple, list)` fits here for array field, but is not general purpose style
- branching and looping pressure is quite high (every element of all containers opens a branch, not sure if there are further all, any or reduce tricks possible to lower the burden by doing faster conditional eval on C side)

Not sure if any of that matters here though.

Reply:
@yetanotherion I basically agree with that, thus here are two more versions:

```python
def flatten_stack(items, seqtypes=(list, tuple)):
    ii_stack = []
    ii = iter(items)
    while True:
        try:
            n = next(ii)
            if isinstance(n, seqtypes):
                ii_stack.append(ii)
                ii = iter(n)
            else:
                yield n
        except StopIteration:
            try:
                ii = ii_stack.pop()
            except IndexError:
                break

def flatten_latestack(items, seqtypes=(list, tuple)):
    for e1 in items:
        if isinstance(e1, seqtypes):
            for e2 in e1:
                if isinstance(e2, seqtypes):
                    ii_stack = []
                    ii = iter(e2)
                    while True:
                        try:
                            n = next(ii)
                            if isinstance(n, seqtypes):
                                ii_stack.append(ii)
                                ii = iter(n)
                            else:
                                yield n
                        except StopIteration:
                            try:
                                ii = ii_stack.pop()
                            except IndexError:
                                break
                else:
                    yield e2
        else:
            yield e1
```
Both versions are recursion-free, `latestack` combines the ideas from `unroll` and `stack` (and is only tiny bit slower than unroll3 for relatively flat structures, but alot faster for deep nested ones).

Reply:
Great! @jerch WDYT about:

```
def flatten_latestack(items, seqtypes=(list, tuple)):
    for e1 in items:
        if isinstance(e1, seqtypes):
            for e2 in e1:
                if isinstance(e2, seqtypes):
                    for e3 in e2:
                        if isinstance(e3, seqtypes):
                            yield from flatten_stack(items,
                                                     seqtypes=seqtypes)
                        else:
                            yield e3
                else:
                    yield e2
        else:
            yield e1
```
The idea being:
- we keep the unroll3 (I would not be surprised that 3 dimensional arrays would be used often enough to benefit from this unrolling).
- if it's more than 3, we pay the price of function calls for the sake of an easier readibility/maintainability?

Reply:
Btw in `flatten_stack` is another tiny speedup possible, if the chances for list|tuple types is rather low - the if/else branching could be turned into a while loop:
```python
...
n = next(ii)
while not isinstance(e3, seqtypes):
    yield e3
    n = next(ii)
ii_stack.append(ii)
ii = iter(n)
...
```

Reply:
Indeed, actually based on these tests, the difference does not seem to be pretty big.

```python
def flatten_stack(items, seqtypes=(list, tuple)):
    ii_stack = []
    ii = iter(items)
    while True:
        try:
            n = next(ii)
            if isinstance(n, seqtypes):
                ii_stack.append(ii)
                ii = iter(n)
            else:
                yield n
        except StopIteration:
            try:
                ii = ii_stack.pop()
            except IndexError:
                break

def flatten_stack_bis(items, seqtypes=(list, tuple)):
    ii_stack = []
    ii = iter(items)
    while True:
        try:
            n = next(ii)
            while not isinstance(n, seqtypes):
                yield n
                n = next(ii)
            ii_stack.append(ii)
            ii = iter(n)
        except StopIteration:
            try:
                ii = ii_stack.pop()
            except IndexError:
                break
```
```python
r = [[[[None
      for _ in range(10)]
      for _ in range(10)]
      for _ in range(10)]
      for _ in range(365)]


r_two =[[None
         for _ in range(1000)]
         for _ in range(365)]



expected = [None for _ in range(10 * 10 * 10 * 365)]
expected_two = [None for _ in range(365 * 1000)]
def test_flatten_stack(benchmark):
  assert expected == benchmark(lambda: list(flatten_stack(r)))

def test_flatten_stack_two(benchmark):
  assert expected_two == benchmark(lambda: list(flatten_stack(r_two)))


def test_flatten_stack_bis(benchmark):
  assert expected == benchmark(lambda: list(flatten_stack_bis(r)))


def test_flatten_stack_bis_two(benchmark):
  assert expected_two == benchmark(lambda: list(flatten_stack_bis(r_two))) 
```
```
Name (time in ms)                  Min                Max               Mean
------------------------------------------------------------------------------------
test_flatten_stack_bis_two     29.3612 (1.0)      29.9591 (1.0)      29.5208 (1.0)
test_flatten_stack_two         30.3869 (1.03)     30.8814 (1.03)     30.5511 (1.03)
test_flatten_stack_bis         39.7383 (1.35)     40.1010 (1.34)     39.8639 (1.35)
test_flatten_stack             40.7907 (1.39)     41.0936 (1.37)     40.9310 (1.39)
```

I'm personally fine with both versions of flatten_stack, I tend to think that the most readable should be chosen here. Do you have a preference?

Reply:
Have no strong preference either, though I find the second variant somewhat easy to grasp...

Reply:
Just pushed a new version. There may be a small inconsistency between `flatten` and `_flatten_stack`:
- if isinstance else, vs
- while not isinstance,

Still reading the code I find it easy to grasp too. Waiting for your inputs @jerch @shangxiao before going further.
(If that's fine with you let's focus on `flatten` first, I'll add the missing `nested nested nested array` in the second commit's test after.)

Reply:
> `if isinstance else`, vs `while not isinstance`

Indeed, I think we should not flip the condition evaluation if it can be avoided. Maybe flip the branching under the for loops, or just revert back to the older `stack` version?